src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs

```diff
@@ -41,93 +50,344 @@ public class AllSubscription : IAsyncEnumerator<ResolvedEvent> {
 				}
 
 				_bus = bus;
-				_nextPosition = position;
 				_resolveLinks = resolveLinks;
 				_user = user;
 				_readIndex = readIndex;
-				_disposedTokenSource = new CancellationTokenSource();
-				_buffer = new ConcurrentQueue<ResolvedEvent>();
-				_tokenRegistration = cancellationToken.Register(_disposedTokenSource.Dispose);
-			}
+				_cancellationToken = cancellationToken;
+				_connectionId = Guid.NewGuid();
 
-			public ValueTask DisposeAsync() {
-				_disposedTokenSource.Dispose();
-				_tokenRegistration.Dispose();
-				return default;
+				_inner = new CatchupAllSubscription(bus, position, resolveLinks, user, readIndex, cancellationToken);
 			}
 
 			public async ValueTask<bool> MoveNextAsync() {
-				ReadLoop:
-				if (_disposedTokenSource.IsCancellationRequested) {
-					return false;
+				var result = await _inner.MoveNextAsync().ConfigureAwait(false);
+				if (result) {
+					return true;
 				}
 
-				if (_buffer.TryDequeue(out var current)) {
-					_current = current;
-					return true;
+				if (!_catchUpRestarted) {
+					await _inner.DisposeAsync().ConfigureAwait(false);
```

Comment:
It doesn't appear that we ever move to a live subscription here unless a cancellation has been requested. Are we missing some logic where we decide to switch over to the live subscription based on `IsEndOfStream` from the `ReadAllEventsForwardCompleted`.